### PR TITLE
feat: new hooks

### PR DESCRIPTION
**Pull request by ronag**

### CODE REVIEW

**Comment by ronag** [comment]
I'm unsure about `context` in `onRequestStart`. 

A. Is it necessary?
B. Should it live on the controller instead?

File: test/mock-agent.js [hunk 1]
```diff
@@ -142,89 +142,6 @@ describe('MockAgent - dispatch', () => {
...
-      method: 'GET'
-    }, {
-      onError: (err) => { throw err },
-      onConnect: () => {},
-      onBodySent: () => {},
-      onHeaders: () => {},
-      onData: () => {},
-      onComplete: 'INVALID'
-    }), new InvalidArgumentError('invalid onComplete method'))
-  })
```
Comment (mcollina) [inline test/mock-agent.js:227]: why was this removed?

**Comment by mcollina** [comment]
This looks good to me. Can you update the docs?

**Review by mcollina — APPROVED** [review]
lgtm

File: test/mock-agent.js (hunk 1 above)
Comment (ronag) [reply test/mock-agent.js:227]: The redirect interceptor has some kind of assertion of the handler that does not exist anywhere else and is difficult to have compat with through the wrap/unwrap.

**Comment by tim1uphealth** [comment]
Nit: could update the docs example, it's currently using the deprecated hooks: https://github.com/nodejs/undici/blob/6bb527e9d768c5ca5b98765c28cf403ad8b250fb/docs/docs/api/Dispatcher.md?plain=1#L239

**Comment by metcoder95** [comment]
Would you like to send a PR to update documentation?
